File: frontend/src/api/types/indexing.ts
export type IndexJobStatus = "pending" | "running" | "completed" | "failed" | "cancelled";

export interface IndexJob {
  id: number;
  collection_id: number;
  status: IndexJobStatus;
  documents_total: number;
  documents_processed: number;
  terms_indexed: number | null;
  error_message: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
}

export const TERMINAL_INDEX_STATUSES: readonly IndexJobStatus[] = ["completed", "failed", "cancelled"];

File: packages/ips_db/ips_db/models/indexing.py
from __future__ import annotations

import datetime as dt

from sqlalchemy import ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base


class IndexJob(Base):
    __tablename__ = "index_jobs"

    id: Mapped[int] = mapped_column(primary_key=True)
    collection_id: Mapped[int] = mapped_column(ForeignKey("collections.id", ondelete="CASCADE"))
    status: Mapped[str] = mapped_column(String(20), default="pending")

    documents_total: Mapped[int] = mapped_column(Integer, default=0)
    documents_processed: Mapped[int] = mapped_column(Integer, default=0)
    terms_indexed: Mapped[int | None] = mapped_column(Integer, nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[dt.datetime] = mapped_column(server_default=func.now())
    started_at: Mapped[dt.datetime | None] = mapped_column(nullable=True)
    finished_at: Mapped[dt.datetime | None] = mapped_column(nullable=True)
